connectd/connect.c

```diff
@@ -1038,68 +1055,36 @@ gossip_resolve_addr(const tal_t *ctx, const struct pubkey *id)
 			      strerror(errno));
 
 	msg = wire_sync_read(tmpctx, GOSSIPCTL_FD);
-	if (!fromwire_gossip_get_addrs_reply(tmpctx, msg, &addrs))
+	if (!fromwire_gossip_get_addrs_reply(tmpctx, msg, &normal_addrs))
 		status_failed(STATUS_FAIL_INTERNAL_ERROR,
 			      "Failed parsing get_addrs_reply gossipctl: %s",
 			      tal_hex(tmpctx, msg));
 
-	if (!addrs)
-		return NULL;
-
-	/* FIXME: Don't just take first address! */
-	addr = tal(ctx, struct wireaddr_internal);
-	addr->itype = ADDR_INTERNAL_WIREADDR;
-	addr->u.wireaddr = addrs[0];
-
-	return addr;
+	/* Wrap each one in a wireaddr_internal and add to addrs. */
+	for (size_t i = 0; i < tal_count(normal_addrs); i++) {
+		struct wireaddr_internal addr;
+		addr.itype = ADDR_INTERNAL_WIREADDR;
+		addr.u.wireaddr = normal_addrs[i];
+		append_addr(addrs, &addr);
+	}
 }
 
-/* Consumes addrhint if not NULL */
-static void try_reach_peer(struct daemon *daemon,
-			   const struct pubkey *id,
-			   u32 seconds_waited,
-			   struct wireaddr_internal *addrhint)
+static void try_reach_one_addr(struct reaching *reach)
 {
-	struct wireaddr_internal *a;
-	int fd, af;
-	struct reaching *reach;
-	bool use_proxy = daemon->use_proxy_always;
+ 	int fd, af;
+	bool use_proxy = reach->daemon->use_proxy_always;
 
-	/* Already done?  May happen with timer. */
-	if (pubkey_set_get(&daemon->peers, id))
-		return;
-
-	/* If we're trying to reach it right now, that's OK. */
-	if (find_reaching(daemon, id))
+	if (reach->addrnum == tal_count(reach->addrs)) {
+		connect_failed(reach->daemon, &reach->id, reach->seconds_waited,
+			       reach->addrhint, "%s", reach->errors);
+		tal_free(reach);
 		return;
-
-	/* FIXME: Gather *all* the addresses, and iterate through. */
-	a = addrhint;
-	if (!a)
-		a = gossip_resolve_addr(tmpctx, id);
-
-	if (!a) {
-		/* Don't resolve via DNS seed if we're supposed to use proxy. */
-		if (use_proxy) {
-			a = tal(tmpctx, struct wireaddr_internal);
-			wireaddr_from_unresolved(a, seedname(tmpctx, id),
-						 DEFAULT_PORT);
-		} else if (daemon->use_dns) {
-			a = seed_resolve_addr(tmpctx, id,
-					      daemon->broken_resolver_response);
-		}
 	}
 
-	if (!a) {
-		connect_failed(daemon, id, seconds_waited, addrhint,
-			       "No address known");
-		return;
-	}
+ 	/* Might not even be able to create eg. IPv6 sockets */
+ 	af = -1;
 
-	/* Might not even be able to create eg. IPv6 sockets */
-	af = -1;
-
-	switch (a->itype) {
+	switch (reach->addrs[reach->addrnum].itype) {
```

Comment:
Nit: Pull out into an `addr` pointer instead of indexing everywhere.

Reply:
Yeah, agreed.